File: projects/enterprise-app/backend/src/main/java/com/enterprise/mastery/service/UserService.java
package com.enterprise.mastery.service;

import com.enterprise.mastery.core.error.exception.BusinessException;
import com.enterprise.mastery.core.error.exception.ResourceNotFoundException;
import com.enterprise.mastery.domain.entity.Role;
import com.enterprise.mastery.domain.entity.User;
import com.enterprise.mastery.domain.repository.UserRepository;
import com.enterprise.mastery.dto.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.StringUtils;

@Slf4j
@Service
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    @Transactional(readOnly = true)
    public PageResponse<UserResponseDto> getUsers(
            String query,
            Role role,
            Boolean active,
            int page,
            int size,
            String sortBy,
            String sortDir
    ) {
        Sort sort = Sort.by(Sort.Direction.fromString(sortDir.toUpperCase()), sortBy);
        Pageable pageable = PageRequest.of(page, size, sort);

        // Fetch page of users
        Page<User> userPage = userRepository.findAll(pageable);

        Page<UserResponseDto> dtoPage = userPage.map(this::mapToDto);
        return PageResponse.from(dtoPage);
    }

    @Transactional(readOnly = true)
    public UserResponseDto getUserById(Long id) {
        return userRepository.findById(id)
                .map(this::mapToDto)
                .orElseThrow(() -> new ResourceNotFoundException("User", id));
    }

    @Transactional
    public UserResponseDto createUser(UserCreateRequest request) {
        if (userRepository.existsByUsername(request.getUsername())) {
            throw new BusinessException("USERNAME_TAKEN", "Username is already taken.");
        }

        if (userRepository.existsByEmail(request.getEmail())) {
            throw new BusinessException("EMAIL_TAKEN", "Email address is already in use.");
        }

        User user = User.builder()
                .username(request.getUsername())
                .email(request.getEmail())
                .passwordHash(passwordEncoder.encode(request.getPassword()))
                .roles(request.getRoles())
                .active(request.isActive())
                .build();

        return mapToDto(userRepository.save(user));
    }

    @Transactional
    public UserResponseDto updateUser(Long id, UserUpdateRequest request) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User", id));

        if (!user.getEmail().equalsIgnoreCase(request.getEmail()) && userRepository.existsByEmail(request.getEmail())) {
            throw new BusinessException("EMAIL_TAKEN", "Email is already taken by another user.");
        }

        user.setEmail(request.getEmail());
        user.setRoles(request.getRoles());
        user.setActive(request.isActive());

        return mapToDto(userRepository.save(user));
    }

    @Transactional
    public void deleteUser(Long id) {
        if (!userRepository.existsById(id)) {
            throw new ResourceNotFoundException("User", id));
        }
        userRepository.deleteById(id);
    }

    @Transactional(readOnly = true)
    public AvailabilityResponse checkUsernameAvailability(String username) {
        boolean exists = StringUtils.hasText(username) && userRepository.existsByUsername(username);
        return AvailabilityResponse.builder()
                .field("username")
                .value(username)
                .available(!exists)
                .build();
    }

    @Transactional(readOnly = true)
    public AvailabilityResponse checkEmailAvailability(String email) {
        boolean exists = StringUtils.hasText(email) && userRepository.existsByEmail(email);
        return AvailabilityResponse.builder()
                .field("email")
                .value(email)
                .available(!exists)
                .build();
    }

    private UserResponseDto mapToDto(User user) {
        return UserResponseDto.builder()
                .id(user.getId())
                .username(user.getUsername())
                .email(user.getEmail())
                .roles(user.getRoles())
                .active(user.isActive())
                .createdAt(user.getCreatedAt())
                .updatedAt(user.getUpdatedAt())
                .build();
    }
}

File: projects/enterprise-app/frontend/src/app/core/services/user.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { AvailabilityResponse, PageResponse, User, UserCreateRequest, UserFilterParams, UserUpdateRequest } from '../models/user.models';

@Injectable({
  providedIn: 'root'
})
export class UserService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = '/api/v1/users';

  getUsers(filters: UserFilterParams = {}): Observable<PageResponse<User>> {
    let params = new HttpParams();

    if (filters.query) params = params.set('query', filters.query);
    if (filters.role) params = params.set('role', filters.role);
    if (filters.active !== undefined) params = params.set('active', filters.active.toString());
    if (filters.page !== undefined) params = params.set('page', filters.page.toString());
    if (filters.size !== undefined) params = params.set('size', filters.size.toString());
    if (filters.sortBy) params = params.set('sortBy', filters.sortBy);
    if (filters.sortDir) params = params.set('sortDir', filters.sortDir);

    return this.http.get<PageResponse<User>>(this.baseUrl, { params });
  }

  getUserById(id: number): Observable<User> {
    return this.http.get<User>(`${this.baseUrl}/${id}`);
  }

  createUser(request: UserCreateRequest): Observable<User> {
    return this.http.post<User>(this.baseUrl, request);
  }

  updateUser(id: number, request: UserUpdateRequest): Observable<User> {
    return this.http.put<User>(`${this.baseUrl}/${id}`, request);
  }

  deleteUser(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  checkUsername(username: string): Observable<AvailabilityResponse> {
    const params = new HttpParams().set('username', username);
    return this.http.get<AvailabilityResponse>(`${this.baseUrl}/check-username`, { params });
  }

  checkEmail(email: string): Observable<AvailabilityResponse> {
    const params = new HttpParams().set('email', email);
    return this.http.get<AvailabilityResponse>(`${this.baseUrl}/check-email`, { params });
  }
}
